src/components/Edit.vue

```diff
@@ -0,0 +1,193 @@
+<script setup lang="ts">
+import { useAsyncState } from '@vueuse/core'
+import { computed, ref } from 'vue'
+import { useApi } from '../composables/useApi'
+import { useAuth } from '../composables/useAuth'
+import { useSessionToken } from '../composables/useSessionToken'
+import { Leaderboards } from '../lib/api/Leaderboards'
+import {
+	ProblemDetails,
+	UpdateLeaderboardRequest
+} from '../lib/api/data-contracts'
+import { onBeforeRouteLeave, useRouter } from 'vue-router'
+import { HttpResponse } from '../lib/api/http-client'
+
+const props = defineProps<{
+	id: number
+}>()
+
+const updateError = ref('')
+
+const token = useSessionToken()
+
+const router = useRouter()
+
+const leaderboards = new Leaderboards({
+	baseUrl: import.meta.env.VITE_BACKEND_URL
+})
+
+const warnBeforeLeave = ref(true)
+
+const updateRequest = ref<UpdateLeaderboardRequest>({
+	info: '',
+	name: '',
+	slug: ''
+})
+
+const {
+	state: board,
+	error,
+	isLoading,
+	execute
+} = useAsyncState(async () => {
+	const resp = await leaderboards.getLeaderboard(props.id)
+	updateRequest.value.info = resp.data.info
+	updateRequest.value.name = resp.data.name
+	updateRequest.value.slug = resp.data.slug
+	return resp.data
+}, null)
+
+onBeforeRouteLeave(() => {
+	if (
+		warnBeforeLeave.value &&
+		(board.value?.name !== updateRequest.value.name ||
+			board.value?.slug !== updateRequest.value.slug ||
+			board.value?.info !== updateRequest.value.info)
+	) {
+		if (!window.confirm('Do you want to leave? You have unsaved changes.')) {
+			return false
+		}
+	}
+})
+
+const errorResponse = computed(
+	() => (error.value as HttpResponse<unknown, ProblemDetails>).error
+)
+
+async function submit() {
+	useApi(
+		() =>
+			leaderboards.updateLeaderboard(
+				props.id,
+				{
+					name: updateRequest.value.name,
+					info: updateRequest.value.info,
+					slug: updateRequest.value.slug
+				},
+				useAuth(token.value)
+			),
+		() => {
+			warnBeforeLeave.value = false
+			router.push({ name: 'leaderboardView', params: { id: props.id } })
+		},
+		(error) => {
+			updateError.value = 'Failed to update: ' + error.status.toString(10)
+		}
+	)
+}
+</script>
+
+<template>
+	<div class="container">
+		<div v-if="isLoading">Loading...</div>
+		<div v-else-if="error" class="error-container">
+			<p class="errorText">
+				<!-- For unexpected server errors; fields will be empty, hence the defaults. -->
+				Failed to fetch leaderboard: {{ errorResponse.status ?? '500' }}
+				{{ errorResponse.title ?? 'Check the console for more info.' }}
```

Comment:
This is a little grody. If you want a guaranteed status code, get rid of the `.error` on line 64 and use the request's status property instead. You'll still need a fallback for the title though.

Reply:
You mean the response's? Also is this change more what you had in mind?
```
Failed to fetch leaderboard: {{ errorResponse.status ?? '500' }}
{{ errorResponse.error.title ?? 'Check the console for more info.' }}
```

Reply:
I meant to say "response," yeah. The whole point is that you shouldn't need a fallback for the status code now since the error response object should be guaranteed to have one, whereas the actual payload may not even exist. You can change the type to something like `HttpResponse<unknown, void | ProblemDetails>` to reflect this if you want.